File: LEETCODE/HARD/DAY 11/LONGEST SUBSTRING (TWO DISTINCT CHARACTERS).java
import java.util.*;

class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int i = 0, maxLen = 0;

        for (int j = 0; j < s.length(); j++) {
            map.put(s.charAt(j), map.getOrDefault(s.charAt(j), 0) + 1);

            while (map.size() > 2) {
                map.put(s.charAt(i), map.get(s.charAt(i)) - 1);
                if (map.get(s.charAt(i)) == 0) {
                    map.remove(s.charAt(i));
                }
                i++;
            }
            maxLen = Math.max(maxLen, j - i + 1);
        }
        return maxLen;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.lengthOfLongestSubstringTwoDistinct("eceba"));   // 3
        System.out.println(sol.lengthOfLongestSubstringTwoDistinct("ccaabbb")); // 5
    }
}

File: LEETCODE/HARD/DAY 11/LONGEST SUBSTRING (TWO DISTINCT CHARACTERS).py
def lengthOfLongestSubstringTwoDistinct(s: str) -> int:
    from collections import defaultdict
    
    count = defaultdict(int)
    i = 0
    max_len = 0
    
    for j, ch in enumerate(s):
        count[ch] += 1
        
        while len(count) > 2:
            count[s[i]] -= 1
            if count[s[i]] == 0:
                del count[s[i]]
            i += 1
        
        max_len = max(max_len, j - i + 1)
    
    return max_len


# 🚀 Example Run
print(lengthOfLongestSubstringTwoDistinct("eceba"))   # Output: 3
print(lengthOfLongestSubstringTwoDistinct("ccaabbb")) # Output: 5
